### ISSUE TITLE
Children-First Traversal Option

### ISSUE THREAD

**Issue by Jamesernator** [babel/babel#4956]
A problem I've had with using babel-traverse is that `replaceWith` will cause infinite recursion if a node of the same type is used in replacement. This happens because traverse visits the parent before it visits the children for example:

![call_expressions](https://cloud.githubusercontent.com/assets/8518303/20985993/ac7f9c3e-bd2b-11e6-9d79-0f7fd0b16785.png)

If I replace CallExpressions with YieldExpressions (I've been trying to convert regular functions into coroutines).

![call_expressions2](https://cloud.githubusercontent.com/assets/8518303/20986091/026f2682-bd2c-11e6-89b1-d173d5fe2eeb.png)

So it'll revisit CallExpression and replace it again:

![call_expressions3](https://cloud.githubusercontent.com/assets/8518303/20986172/5f956fba-bd2c-11e6-94ff-b83d4a080366.png)

Etc, etc

Now this problem is identical to the list replacement problem e.g.

```javascript
const a = [1, 2, 1, 2]
let i = 0
// Setting it to the original length is a non-solution as well
while (i < a.length ) {
    if (a[i] == 2) {
        a.splice(i, 0, 2)
    }
    i += 1
}

console.log(a) // Never reaches, MemoryError before
```

The solution to the list replacement problem: Iterate Backwards:

```javascript
const a = [1,2,1,2]

let i = a.length
while (i > 0) {
    i -= 1
    if (a[i] === 2) {
        a.splice(i, 0, 2)
    }
}

console.log(a) // [1, 2, 2, 1, 2, 2]
```

For babel-traverse this is fixable by simply adding a children-first traverse (where every child node is traversed before its parent). I'd imagine it being something like:

```javascript
traverseUp(someAST, someVisitor, ...)
```
#### Comments

While I'm aware solutions exist such as adding properties to nodes (https://github.com/thejameskyle/babel-handbook/issues/94) this approach becomes quickly verbose if replaced with more complicated expressions that can't simply check parent (e.g. nested expressions, or if *other* transforms need to happen before trying to check parents)

**Comment by hzoo** [comment]
Wow this is an amazing issue description @Jamesernator!

I totally get where you're coming at: `replaceWith` causing an infinite loop definitely is a concern when learning to write plugins and the workaround for property checking really sucks.

**Comment by jridgewell** [comment]
I've taken to using a `WeakSet` to store replaced nodes. http://astexplorer.net/#/oCVD5qd7Nm

**Comment by Jamesernator** [comment]
@jridgewell While solvable my complete set of use cases is actually a bit more of a pain than just using a WeakSet as it actually adds additional callExpressions into the yield expression:

E.G. I'm not just wanting to wrap CallExpressions in a YieldExpression I want to do something like this:

```javascript
function someStudentFunction(name) {
    let hello = "Hello"
    return hello + name
}

function greeting() {
    let greeting = someStudentFunction("Alice")
    console.log(greeting)
}
```

Will turn into something like:

```javascript
const transformed = Symbol()

function* resolveFunction(func, ...args) {
    // If the function has been transformed yield to it otherwise just
    // return the value
    if (func[transformed]) {
        return yield* func(...args)
    } else {
        return func(...args)
    }
}

function someStudentFunction(name) {
    const wrapped = function*() {
        let hello = (yield { sourceLoc, ...etc }, "Hello")
        return (yield, (yield, hello) + (yield, name))
    }
    wrapped[transformed] = true
    return wrapped
}

function greeting() {
    const wrapped = function*() {
        let greet = yield* resolveFunction(someStudentFunction, "Alice") // new CallExpression here
        yield* resolveFunction(console.log, greet)
    }
    wrapped[transformed] = true
    return wrapped
}
```

**Comment by jridgewell** [comment]
Hmm. You can also [`path.skip`](http://astexplorer.net/#/efhJqnJv5a/1) whatever you replace if you don't need to traverse down again (to transform some other value you've just inserted).

**Comment by Jamesernator** [comment]
Inner calls may still need to be transformed though e.g.

```javascript

function square() ...

function foo() {
    console.log(square(10))
}
```

would need to become

```javascript
...
function foo() {
    const wrapped = function*() {
        // inner call square(10) needs to be reached we can't skip the inner stuff
        yield* resolveFunction(console.log, yield* resolveFunction(square, 10))
    }
    ...
}
...
```

A viable solution for my use case would also be the ability to traverse only a specific subtree, for example:

```javascript
const visitor = {
    CallExpression(path) {
        path.replaceWith(
            template(`
                yield* RESOLVER_FUNCTION(FUNCTION, ARGUMENTS)
            `)({
                RESOLVER_FUNCTION: resolverFunctionID, // generated elsewhere,
                FUNCTION: path.node.callee,
                ARGUMENTS: t.spreadElement(path.node.arguments)
            })
        )
        path.traverseSubtree(path // traverse only the part of the tree I need but not resolveFunction(...) itself
            .node
            .argument // of the yield ...
            .arguments // of the RESOLVER_FUNCTION(...)
        )
        // AND may need to do it for FUNCTION as well if there's a computed
        // function or something
        path.skip()
    }
}

```

**Comment by jridgewell** [comment]
To answer the first part, you can transform in the exit call. http://astexplorer.net/#/rsTjUbjZC1/1

As for you `traverseSubtree`, you can do that, too:

```js
path.replace(...);
path.skip();
path.get("argument.arguments").traverse(visitor);
```

**Comment by Jamesernator** [comment]
I was not aware of `enter()/exit()` that solves basically everything I need. I'll still leave this issue open but that info gives a pretty simple solution for traverseUp, it'd essentially just be `traverse` but treating visitors that are a single function as `exit()` instead of `enter()`

**Comment by hzoo** [comment]
We could add some clarifying info to https://github.com/thejameskyle/babel-handbook/blob/master/translations/en/plugin-handbook.md about this

**Comment by Jamesernator** [comment]
Oh also `traverseUp` would need to add `path.skip()` after each `exit()` implicity (so that it doesn't retraverse the node when replaced).

As for the handbook it'd be nice to have a tiny bit that explains that if you're replacing a node with a child of the same type that you should probably look into `exit()` visitors.

**Comment by hzoo** [comment]
Would you like to make the PR @Jamesernator?

**Comment by chilloutman** [comment]
I recently can into this and would like to provide an example:

Let's say we want to simplify this expression, by which I mean reduce it by folding all constants and statically knowable subexpressions, resulting in "3":

```
((a, b) => a + b)(1, 2)
```

The visitor get's called in this order:
```
enter: CallExpression
enter: ArrowFunctionExpression
enter: BinaryExpression:  a + b
enter: Identifier: a
exit:  Identifier: a
enter: Identifier: b
exit:  Identifier: b
exit:  BinaryExpression: a + b
exit:  ArrowFunctionExpression
enter: NumericLiteral: 1
exit:  NumericLiteral: 1
enter: NumericLiteral: 2
exit:  NumericLiteral: 2
exit:  CallExpression
```

Because we get to the `arguments` only after we exited the `ArrowFunctionExpression`, there is no way of storing the argument values for use inside the function `body`.

My current solution is to use sub-traversal when entering the `ArrowFunctionExpression` to evaluate the `arguments` first. A naive idea would be to reverse it so that the arguments are visited before the body.

(I'm working on de-obfuscation, so this might not be useful for general use cases.)
